### ISSUE TITLE
3.1.0 server seems incompatible with 2.5.2 clients

### ISSUE THREAD

**Issue by vt-alt** [grke/burp#906]
Change introduced in v3 seems to be minor as we never used protocol 2 anyway, and as we use Linux we should not be affected by Windows functions.
```
2022-05-01 burp-3.1.0
  * Remove protocol 2:
  * Remove some unused Windows functions.
2021-08-31 burp-2.5.4
...
```
But after upgrading the server some clients starting to fail to backup. Example error log:
```
0000281 2022-07-08 21:41:59 +0300

Last 1000 lines of /tank/burp/debuginfo.mskdc.altlinux.org/working/log:

2022-07-08 21:41:59 +0300: burp[500537] Backup started
2022-07-08 21:41:59 +0300: burp[500537] Client version: 2.5.2
2022-07-08 21:41:59 +0300: burp[500537] Using librsync hash unset
2022-07-08 21:41:59 +0300: burp[500537] WARNING: Client 'debuginfo.mskdc.altlinux.org' version '2.5.2' does not match server version '3.1.0'. An upgrade is
recommended.
2022-07-08 21:41:59 +0300: burp[500537] Begin phase1 (file system scan)
2022-07-08 21:41:59 +0300: burp[500537] WARNING: Will not descend: file system change not allowed /archive: No such file or directory
2022-07-08 21:42:00 +0300: burp[500537] WARNING: Will not descend: file system change not allowed /proc: No such file or directory
2022-07-08 21:42:01 +0300: burp[500537] End phase1 (file system scan)
2022-07-08 21:42:01 +0300: burp[500537] Begin phase2 (receive file data)
2022-07-08 21:42:01 +0300: burp[500537] main socket 5: Peer closed SSL session
2022-07-08 21:42:01 +0300: burp[500537] error in backup_phase2_server_protocol1
2022-07-08 21:42:01 +0300: burp[500537]   last tried file:    f:000D:/bin/basename
2022-07-08 21:42:01 +0300: burp[500537] End phase2 (receive file data)
2022-07-08 21:42:01 +0300: burp[500537] error in backup phase 2
2022-07-08 21:42:01 +0300: burp[500537] Backup failed
```
`backup_phase2_server_protocol1` is not too informative error message. I thought maybe client is decided to fail and not informing the server, so client log: (Other host with older burp, since I cleared scroll and upgraded client there (and it started to work), but error message is similar):
```
2022-07-08 21:55:21 +0300: burp[976216] Server version: 3.1.0
2022-07-08 21:55:21 +0300: burp[976216] nocsr ok
2022-07-08 21:55:21 +0300: burp[976216] SSL is using cipher: TLS_AES_256_GCM_SHA384  TLSv1.3 Kx=any      Au=any  Enc=AESGCM(256) Mac=AEAD

2022-07-08 21:55:21 +0300: burp[976216] extra_comms_begin ok:autoupgrade:incexc:orig_client:uname:failover:vss_restore:regex_icase:counters_json:msg:rshash=blake2:seed:
2022-07-08 21:55:21 +0300: burp[976216] Compression level: 4
2022-07-08 21:55:21 +0300: burp[976216] do backup client
2022-07-08 21:55:21 +0300: burp[976216] Phase 1 begin (file system scan)
2022-07-08 21:55:21 +0300: burp[976216] WARNING: Will not descend: file system change not allowed /proc: No such file or directory
2022-07-08 21:55:22 +0300: burp[976216] WARNING: Will not descend: file system change not allowed /var/lib/machines: No such file or directory
2022-07-08 21:55:22 +0300: burp[976216] Phase 1 end (file system scan)
2022-07-08 21:55:22 +0300: burp[976216] Phase 2 begin (send backup data)

2022-07-08 21:55:22 +0300: burp[976216] unexpected command in deal_with_read(): t:000E:t/bin/basename
2022-07-08 21:55:22 +0300: burp[976216] Error in backup
2022-07-08 21:55:22 +0300: burp[976216] End backup
```
Maybe this is unrelated to upgrade though! The server is experienced disk full on burp storage partition before upgrading. But upgrading client to 3.1.0 solved problem for particular client. Some other clients still failing and I don't have access to upgrade them - is there a way to fix backup?

**Comment by vt-alt** [comment]
When I set `librsync = 0` on the server, on client:
```
# burp -at
2022-07-08 23:16:01 +0300: burp[979636] Privileges switched to _burp keeping readall capability.
2022-07-08 23:16:01 +0300: burp[979636] Connecting to amfora:4971
2022-07-08 23:16:01 +0300: burp[979636] WARNING: w:0076:Client 'caltrop' version '2.3.22' does not match server version '3.1.0'. An upgrade is recommended.

2022-07-08 23:16:01 +0300: burp[979636] auth ok
2022-07-08 23:16:01 +0300: burp[979636] Server version: 3.1.0
2022-07-08 23:16:01 +0300: burp[979636] nocsr ok
2022-07-08 23:16:01 +0300: burp[979636] SSL is using cipher: TLS_AES_256_GCM_SHA384  TLSv1.3 Kx=any      Au=any  Enc=AESGCM(256) Mac=AEAD

2022-07-08 23:16:01 +0300: burp[979636] extra_comms_begin ok:autoupgrade:incexc:orig_client:uname:failover:vss_restore:regex_icase:counters_json:msg:rshash=blake2:seed:
2022-07-08 23:16:01 +0300: burp[979636] Compression level: 4
2022-07-08 23:16:01 +0300: burp[979636] do backup client
2022-07-08 23:16:01 +0300: burp[979636] Phase 1 begin (file system scan)
2022-07-08 23:16:02 +0300: burp[979636] WARNING: Will not descend: file system change not allowed /proc: No such file or directory
2022-07-08 23:16:02 +0300: burp[979636] WARNING: Will not descend: file system change not allowed /var/lib/machines: No such file or directory
2022-07-08 23:16:02 +0300: burp[979636] Phase 1 end (file system scan)
2022-07-08 23:16:02 +0300: burp[979636] Phase 2 begin (send backup data)

2022-07-08 23:16:02 +0300: burp[979636] unexpected command in deal_with_read(): r:003D:A E A gC LgAM IHt B A A A Imo BAA BI Bg8xdh BdFJqC BgEjHC A A
2022-07-08 23:16:02 +0300: burp[979636] Error in backup
2022-07-08 23:16:02 +0300: burp[979636] End backup
```
When I remove `librsync = 0` on the server, error on client:
```
# burp -at
2022-07-08 23:16:20 +0300: burp[979658] Privileges switched to _burp keeping readall capability.
2022-07-08 23:16:20 +0300: burp[979658] Connecting to amfora:4971
2022-07-08 23:16:20 +0300: burp[979658] WARNING: w:0076:Client 'caltrop' version '2.3.22' does not match server version '3.1.0'. An upgrade is recommended.

2022-07-08 23:16:20 +0300: burp[979658] auth ok
2022-07-08 23:16:20 +0300: burp[979658] Server version: 3.1.0
2022-07-08 23:16:20 +0300: burp[979658] nocsr ok
2022-07-08 23:16:20 +0300: burp[979658] SSL is using cipher: TLS_AES_256_GCM_SHA384  TLSv1.3 Kx=any      Au=any  Enc=AESGCM(256) Mac=AEAD

2022-07-08 23:16:20 +0300: burp[979658] extra_comms_begin ok:autoupgrade:incexc:orig_client:uname:failover:vss_restore:regex_icase:counters_json:msg:rshash=blake2:seed:
2022-07-08 23:16:20 +0300: burp[979658] Compression level: 4
2022-07-08 23:16:20 +0300: burp[979658] do backup client
2022-07-08 23:16:20 +0300: burp[979658] Phase 1 begin (file system scan)
2022-07-08 23:16:21 +0300: burp[979658] WARNING: Will not descend: file system change not allowed /proc: No such file or directory
2022-07-08 23:16:21 +0300: burp[979658] WARNING: Will not descend: file system change not allowed /var/lib/machines: No such file or directory
2022-07-08 23:16:21 +0300: burp[979658] Phase 1 end (file system scan)
2022-07-08 23:16:21 +0300: burp[979658] Phase 2 begin (send backup data)

2022-07-08 23:16:22 +0300: burp[979658] unexpected command in deal_with_read(): t:000E:t/bin/basename
2022-07-08 23:16:22 +0300: burp[979658] Error in backup
2022-07-08 23:16:22 +0300: burp[979658] End backup
```
From journalctl on the server:
```
Jul 08 23:16:22 amfora burp[831184]: main socket 5: network write problem in asfd_do_write_ssl: 1 - 0=Success
```

**Comment by vt-alt** [comment]
> But after upgrading the server some clients starting to fail to backup.

Update: All clients v2->v3 are failed to backup. I think this is major issue and it's unexpected. I think we'll downgrade package for ALT Linux to 2.5.4, this is unfortunate.

**Comment by vt-alt** [comment]
FYI. Downgrade of server to 2.5.4 fixed backups:
```
2022-07-09 18:54:44 +0300: burp[1019478] Phase 1 end (file system scan)
2022-07-09 18:54:44 +0300: burp[1019478] Phase 2 begin (send backup data)

2022-07-09 18:55:16 +0300: burp[1019478] Phase 2 end (send file data)
2022-07-09 18:55:16 +0300: burp[1019478] backup finished ok
```
So this is surely version related and not (previous) disk full related.

**Comment by grke** [comment]
Hello,

Yes, there is a fix for this in current master.
I haven't done a versioned release of it yet because I am needing to fix (unrelated) openssl deprecation issues.

**Comment by vt-alt** [comment]
Thanks! OpenSSL deprecation, perhaps, is just a compilation warnings, but this is user visible major bug (because it makes backups impossible — one main thing the program does). More you delay more people will stumble upon it. JFYI distributions update log https://repology.org/project/burp/history

**Comment by grke** [comment]
Hello,

The openssl thing is not just compilation warnings, it's threatening to get burp removed from Debian.

As always, I separate the burp releases into "latest", "stable", and "old stable":
3.1.0: Latest
2.4.0: Stable (recommended)
2.2.18: Old stable
It is recommended to use the current stable version.
The latest release contains new features, but may have unexpected issues.

**Comment by vt-alt** [comment]
IC (We didn't switched to openssl-3 yet).

Yes I'm incorrectly took `3.1.0` as somewhat stablish (since its description was quite innocent). And then misinterpreted this bug (month ago) and pushed 3.1.0 into the stable branch (p10), while testing didn't catch the regression.

Thanks for clarification.

**Comment by grke** [comment]
No worries. :)

**Comment by vt-alt** [comment]
@grke I am curious, is 3.2.0 compatible with 2.5.4 clients?

**Comment by grke** [comment]
Hello,
3.2.0 was built from git master at around Christmas, so it will have the fixes that I was mentioning back in 2022.
It looks like the fix went into 3.1.2:

```
2022-09-01 burp-3.1.2
  * 904: openssl3 support
    - And switch to aes-cbc instead of blowfish, which is deprecated, for
      file encryption
  * 908: Only glibc supports %z in strptime()
  * Fix 2.x clients talking to 3.x servers.
  * Delete heaps more unused code
```
